### ISSUE TITLE
[49.16 - J17] Princess Issues with Advanced Ammo Selection (MMLs and LACs at least).

### ISSUE THREAD

**Issue by Thom293** [MegaMek/megamek#4952]
Dont know how hard this is to take on, but, would be a good boost to her if fixed.

TLDR:  My guess is she is not doing whatever the "what is the best ammo for this round" check until after her first shot with MMLs.  She may not do it ever with normal AC ammo (appears to just default to first in list).  She does appear to make an ammo decision with LBX and swap ammo prior to first firiring and as apprpriate during match.

MMLs: She will not change from starting ammo until shots are fired.  Once she shoots any ammo, she will swap them as needed, but she will not swap from starting ammo until she has shot it - which if a problem is SRM ammo is initially loaded.

Scenario - load two identical MML units in opposition.  Both with SRM and LRM ammo.  SRM ammo initially loaded.  

Neither unit will fire until in SRM range.

Change one to LRM ammo initially loaded, that one will fire in LRM range and then swap to SRM ammo as it closes.  Other unit will still only fire in SRM range.

Add reinforcements 8 rounds later.  After initial battle with both starting units using SRMs, survivor will swap to LRMs and shoot at reinforcements.

So my guess is she is not doing whatever the "what is the best ammo for this round" check until after her first shot.  

Screen 1 shows both in LRM range but not shooting. 
![image](https://github.com/MegaMek/megamek/assets/23645569/01bf6206-26fa-49cd-912a-21faa2795a13)

LAC:  Testing shows LAC-5 equipped mech has never switched from initially loaded basic ammo to either Precision or AP (has both) ammo (with a TN of 4 might want AP ammo, although it didnt matter here).   While approaching should have deffo switched to Precision. It LOOKs like with LAC she will fire whatever is in the first ammo slot, regardless of what is stated to be loaded at the bottom (here they match but did not in most other cases and it it defaulted to first in list).

![image](https://github.com/MegaMek/megamek/assets/23645569/48383111-3f71-472f-9046-f1542ccca077)

LBX:  Screen 3 shows that LBX5 loaded with cluster swithed to solid ammo before first shots.  Also when enemy mech was down with TN of 3, she switched back to cluster even when she still had solid ammo to try to "finish it off."  So, looks like LBX checks for something and checks it pretty well, but MML and LAC do not.

![image](https://github.com/MegaMek/megamek/assets/23645569/4f6a819d-7403-47b2-8b51-1789eedceb1b)

I dont have a screen and dont have time to test right now, but I seem to recall seeing this behavior with hot loaded LRMS before too.  I would assume AtMs and anything with alternate ammos may need to be looked at.  And although not exactly on point, If it matters, if the optional "rapid fire AC" rule is checked, she does seem to make some decision there, as she doesnt always shoot in rapid mode, but will sometimes.

Here are sample saves for both MML and LAC scenarios, game logs and customs for the LAC issue.

[Ammo.zip](https://github.com/MegaMek/megamek/files/13653067/Ammo.zip)

**Comment by Sleet01** [comment]
I haven't looked at this in a while, but I'd guess that the issue is in the two-pass planning code:

1. "guesstimate" stage with fast approximations for most weapon types, to quickly exclude a lot of suboptimal moves / attacks.
2. "accurate" stage which uses the actual to-hit and damage code to refine planning of remaining options.

If #1 doesn't account for various alternative ammo types, it may prematurely discard attacks with weapon systems that don't currently have range (because of the default loaded ammo) without running through the various options.

If that's the case, one possible approach would be to add one guesstimate check for each available ammo type per switchable weapon; this shouldn't increase initial calculation time much, and may allow the existing accurate planning code to then properly determine the correct ammo to use.

If the issue isn't there, in stage #1, then it will probably take a deep dive to figure out where the logic issue is.

**Comment by Thom293** [comment]
yeah Id be interested to see her logic what logic she already uses to choose rapid fire or whether to use solid or Cluster LBX.  LBX cluster gives a -1 TN, but she swaps away from it at long range.  But she does swap back to it when she thinks something is vulnerable.  

So - at least in ammo selection:
1.  does she consider TN at all?  (e.g. IF TN < X, then use AP ammo, IF TMM > Y, use Precision Ammo, etc, if Range = Impossible or Long, Check for LRM ammo, etc)
2. does she consider target properties? (e.g. IF vehicle then Cluster, or Inferno ammo, if a limb stripped, then Cluster ammo or LRM SRM weapon over alternate (if Heat requires a choice)

Can get complicated quickly.  But if she could even do just the items in #1 would be a good improvement.

**Comment by Sleet01** [comment]
Okay, I've taken a quick look and it appears that there are a couple issues:

1.  The fast `guessToHitModifierForWeapon()` has no concept of alternative munitions and doesn't attempt to check anything except the currently linked bin for ammo-using weapons.  This accounts for e.g. not even attempting to fire MMLs with LRM shots if at LRM range but SRMs are loaded: to Princess, this weapon simply cannot hit and is discarded from consideration.
2. In `getPreferredAmmo()`, specific target/target state/munition combos are considered _after_ Princess has decided to attack a target with a weapon, but _only_ specific combos; anything not specifically checked is left on the current ammo. N.B.: this is where the LB-X Cluster/Solid switching occurs, and where the switch _from_ SRM _to_ LRM ammo occurs for MMLs - but the reverse is impossible, because an MML loaded with SRMs has already been skipped (see #1).  Furthermore, this method does not consider what ammo is currently loaded in a given weapon.

I think there are a couple of options here, neither especially savory.  For both, we'll probably want to refactor all of the `buildWeaponFireInfo()` calls in `getFullFiringPlan()` and the like accept a passed-in ammo Mountable, and iterate over both weapons _and_ valid ammo options.  For energy / CI / BA weapons, pass `null` (I'm not sure the latter two even have the option).  Then either:

A)  expand `getPreferredAmmo()` to offer more alt ammo selections for attacks that are chosen.

B)  return back a list of selected ammunition as part of the weaponFireInfo or in AmmoConservation or something, so that we can force the use of the ammo with the highest to-hit options (at least in cases where a specific behavior has not been defined in `getPreferredAmmo()`.

A) has the advantage of needing less refactoring, but will produce unpredictable behavior if, for instance, a given ammo provides the best to-hit and/or damage, but has not been defined in `getPreferredAmmo` - in this case Princess may fire ammo weapons using ammunition that will 100% miss.

B) has the advantage of streamlining ammo selection and not getting screwed up if a given munition hasn't been explicitly added to `getPreferredAmmo` or one of the functions it calls.  I'm not sure there's a disadvantage.

I think we probably should refactor all this code to make a "SwitchAmmoPlan" part of the FiringPlan object, so these two sets of decisions are more organically related, but this would be a huge amount of work.

**Comment by Thom293** [comment]
Thank you.

So it sounds like a lot of work, which is beyond my abilities.  

If you are planning on tackling it, that is awesome.  

But if not, thinking out loud, what if each turn (and presumably at game start, she does a check at the start of her "fast" analysis:  If no enemy in range of current ammo, Switch to ammo with longest range or, failing a range difference, best to-hit mod.  Then,  that will give her the best chance for some firing solution or best odds of one when she is finally in range of something.  And then it more likely enables "get preferred" to do what it already does and do the heavy lifting when possible?  

Certainly there are better ways to do it, but if it's too much to tackle perhaps that would be a bandaid measure that would cover most use cases?  

Again just thinking out loud.  Thanks for looking at this.

**Comment by SuperStucco** [comment]
You don't necessarily want to select the ammo with the best to-hit numbers.  For example, a target with no armor in one or more locations may be better served with SRM ammo from an MML than LRM ammo even though the target number may be 1 or 2 points worse.  Same with selecting SRM inferno ammo vs LRM or standard SRM ammo when shooting at vehicles.  Gets even further into the complicated part if you want to throw in checking for NARCd targets, or guided missiles vs targets with ECM/ECM coverage.  Another example is attacking fixed wing aircraft - while the flak bonus will make it easier to hit, an LBX slug or standard AC round may have a better chance of thresholding the fighters armor while changing the to-hit from 4 to 6 which is not horrible.

This got really tangled when I was getting ye olde TestBot to run through ammo types via brute force iteration of all appropriate bins.  It might be an idea to have static matrices of target types vs weapon ammo types, iterating through the preferred types and checking subsequent types for changes in to-hit number and/or effective damage.  In the above example, the vehicle target progression for SRMs may be Inferno => NARC => Artemis => Standard; first check to see if infernos are available, if not then NARC and what's the difference in to-hit (none in this case, but the process would be universal) and expected damage including penetrating hits against open armor or by ammo type.  If NARC is better then check against Artemis (none carried, so next), and then Standard (carried, but in this specific example less damage and fewer hits because the target has a NARC pod attached).

**Comment by Sleet01** [comment]
I'm planning to fix this up, just wanted to post my findings and ideas so folks could comment before I get started.

> what if each turn (and presumably at game start, she does a check at the start of her "fast" analysis: If no enemy in range of current ammo, Switch to ammo with longest range or, failing a range difference, best to-hit mod. Then, that will give her the best chance for some firing solution or best odds of one when she is finally in range of something. And then it more likely enables "get preferred" to do what it already does and do the heavy lifting when possible?

Yeah, this is pretty much what I'm thinking.  Just checking all the valid ammo options up front will help with stuff like MMLs, probably ATMs too, and should work with the existing special-case ammo handling.  I think putting the "best" ammo in the `weaponFireInfo` class will work fine for paying that info around, and won't require a lot of extra work.

**Comment by Sleet01** [comment]
> You don't necessarily want to select the ammo with the best to-hit numbers. For example, a target with no armor in one or more locations may be better served with SRM ammo from an MML than LRM ammo even though the target number may be 1 or 2 points worse. Same with selecting SRM inferno ammo vs LRM or standard SRM ammo when shooting at vehicles. Gets even further into the complicated part if you want to throw in checking for NARCd targets, or guided missiles vs targets with ECM/ECM coverage. Another example is attacking fixed wing aircraft - while the flak bonus will make it easier to hit, an LBX slug or standard AC round may have a better chance of thresholding the fighters armor while changing the to-hit from 4 to 6 which is not horrible.

This is what `getPreferredAmmo` does now, and is why (for instance) Princess will switch from Solid to Cluster LB-X when an enemy has lost a lot of armor, or switch _back_ from LRM to SRM ammo for MMLs (but not vice-versa).  It's actually pretty effective for ammo it knows about.

> This got really tangled when I was getting ye olde TestBot to run through ammo types via brute force iteration of all appropriate bins. It might be an idea to have static matrices of target types vs weapon ammo types, iterating through the preferred types and checking subsequent types for changes in to-hit number and/or effective damage. In the above example, the vehicle target progression for SRMs may be Inferno => NARC => Artemis => Standard; first check to see if infernos are available, if not then NARC and what's the difference in to-hit (none in this case, but the process would be universal) and expected damage including penetrating hits against open armor or by ammo type. If NARC is better then check against Artemis (none carried, so next), and then Standard (carried, but in this specific example less damage and fewer hits because the target has a NARC pod attached).

I _think_ a lot of this code already exists, in the form of various functions for determining the best available ammo for certain weapon types, but it's not comprehensive.

We might want to consider making the ammo selector a class, containing all the above code; then we could make Princess's ability to select appropriate ammo a tunable setting.  I'll try to keep that option open as I work on this.

**Comment by Thom293** [comment]
> You don't necessarily want to select the ammo with the best to-hit numbers. For example, a target with no armor in one or more locations may be better served with SRM ammo from an MML than LRM ammo even though the target number may be 1 or 2 points worse. Same with selecting SRM inferno ammo vs LRM or standard SRM ammo when shooting at vehicles. Gets even further into the complicated part if you want to throw in checking for NARCd targets, or guided missiles vs targets with ECM/ECM coverage. Another example is attacking fixed wing aircraft - while the flak bonus will make it easier to hit, an LBX slug or standard AC round may have a better chance of thresholding the fighters armor while changing the to-hit from 4 to 6 which is not horrible.
> 


Yeah, sorry.  Didn't mean to load long range as final solution.  Just though from his explanation it's needed to "engage" the second part of the logic where it actually selects the proper ammo.

**Comment by Thom293** [comment]
This looks great.  Did you test ADAs, since that was the issue on your force in the design challenge?

**Comment by Sleet01** [comment]
> This looks great. Did you test ADAs, since that was the issue on your force in the design challenge?

Not yet.  The issue there may be related to ammo ordering, or target availability during the indirect fire phase - but I do want to get that sorted as well.

I think I'm going to need to define a bunch of unit tests to verify all the easy choices - MMLs to LRM ammo, LB-X to cluster at range, etc. - then look at more corner-case stuff.

**Comment by Thom293** [comment]
let me know if you need help testing.
